src/server/run.rs

```diff
@@ -0,0 +1,26 @@
+use std::net::TcpListener;
+
+use actix_web::{dev::Server, web::Data, App, HttpServer};
+use anyhow::{Context, Result};
+use log::info;
+
+use crate::server::router;
+use crate::state::State;
+
+pub fn run(listener: TcpListener) -> Result<Server> {
+    let state = State::new().context("Failed to initialze \"State\"!")?;
+    let state = Data::new(state);
+
+    info!("Serving on http://{}", listener.local_addr()?);
+
+    let server = HttpServer::new(move || {
+        // `Data<State>` is being cloned to avoid creating
+        // one `State` instance per worker.
+        // Otherwise, `Data<State>` will be intialized as many
+        // times as the total count of workers configured
+        App::new().app_data(state.clone()).configure(router)
+    })
+    .listen(listener)?
+    .run();
+    Ok(server)
```

Comment:
Whats the benefit on providing a `TcpListener` ourselves? Is there anything we get by doing this approach?

Actix just needs the desired socket to bind to, this is not something you usually do, using a `SocketAddr` is just enough to achieve this. If we wanted to do something else with this TCP connection apart from using it with Actix perhaps this would help, but at least for now theres no need to consume the same TCP stream as Actix.

If validation is you goal using `SocketAddr` here will do the job, Actix is also capable of receiving this `SocketAddr` instance and bind the server to the corresponding host as you expect.